Include downloading of XBRL zip data and return FilingMetaData.

jreed1701 opened the pull request:
Modify _utils.py to for the url and filename for the XBRL zip
file that comes with 10-K, 10-K/A, 10-Q, 10-Q/A filings.

Modify Downloader to download zip file in addition to the full
text submission file.

Added the kwarg return_meta to the get function. If it's set
the returned value is a tuple with the FilingMetaData object
provided in addition to the filing count.

Limtations:
Should work with document types mentioned above, but could
break for other document types. Also, XBRL did not start
showing up in filing until about 2011-(ish) so it would
make sense to place guards on document type and or make
the XBRL detail files an option. Open to suggestion.

Possible Future work:
This added file is a zip file, could add a feature to unzip it.

Relates to #32 because the FilingMetaData object is returned,
optionally. Believe this is the best option because the files
can be downloaded and the data returned in one web request
instead of two.

jadchaar commented on sec_edgar_downloader/Downloader.py:
```diff
@@ -52,6 +52,7 @@ def get(
         download_details: bool = True,
+        return_meta: bool = False,
         query: str = "",
     ) -> int:
```
Need to change the return type to be `Union[int, Tuple(int, FilingMetadata)]` to match the docstring.

jreed1701 replied:
If return_meta is True, the return type is  ```Union[int, Tuple(int, FilingMetadata)]```  except it's a list of FilingMetaData objects. If that input is false, then it's an int. 

In retrospect, having different return type depending on a switch is probably a bad idea.  What if the return was a list of the meta data objects, and then the user could get the len() of the list?

jadchaar replied:
Yeah I think that might be a good solution. That would break the current API, so this would need to make it into a 5.0 release. No problem, just something to note :)

jreed1701 replied:
Alternatively, I could wrap the list in an object and then the user could do something to the tune of "result.length". Sounds like you're on board with having the function return a single variable/type. I will make this change.

jadchaar commented on sec_edgar_downloader/Downloader.py:
```diff
@@ -171,8 +175,9 @@ def get(
             filing,
             filings_to_fetch,
-            download_details,
+            download_details
```
I think this will cause linting on CI to fail.

jreed1701 replied:
I'll correct it. Wanted to wait for you input before making it pretty.

jadchaar commented on sec_edgar_downloader/_utils.py:
```diff
@@ -300,6 +307,23 @@ def download_filings(
                     )
 
+                try:
+                    download_and_save_filing(
```
`download_and_save_filing` needs to be modified to actually download the zip folder. In its current state, nothing will be downloaded.

jreed1701 replied:
Please clarify? I tested this and it does appear to work correctly.

jadchaar replied:
Oh my bad, I totally misread this hah! I see what you are doing and this totally makes sense :).

jreed1701 replied:
Cool. I am going to mark this conversation resolved.

jadchaar reviewed (changes requested):
Hi @jreed1701 thanks for the contribution! It seems like a number of modifications need to be made to the PR to get it fully functioning. Once those are complete, mind adding some test cases to test the functionality?

jreed1701 commented:
I can attempt to make additional pytests once over the main hurdles.
I don't want the filing-submission.txt so I am putting in another switch like you had for the details.
Since I did that, I also put one in there for the XBRL data too. Gives the user the option for all three.
I did make it so the defaults still exhibit the same behavior I found the code in to begin with.

jadchaar commented:
Mind adding some test cases to your updated and added code (so we can ensure it is functioning as intended)? Then I can give it a thorough review.

jreed1701 commented:
> Mind adding some test cases to your updated and added code (so we can ensure it is functioning as intended)? Then I can give it a thorough review.

Converted this PR to a draft pending those tests.  I will add them as I have time to do so.  Need to figure out TOX first.  I have some outstanding questions of you in prior comments. Please address them at your next convenience.

jadchaar commented:
> > Mind adding some test cases to your updated and added code (so we can ensure it is functioning as intended)? Then I can give it a thorough review.
> 
> Converted this PR to a draft pending those tests. I will add them as I have time to do so. Need to figure out TOX first. I have some outstanding questions of you in prior comments. Please address them at your next convenience.

Ah apologies, things have been hectic on my end so I totally missed that. I will address them now :). Also if you are not comfortable with tox, feel free to just invoke pytest manually like `pytest -v tests`. Tox is just great for creating isolated environments for each test run, so you can run it like this: `tox`. If you just want to target a single python version run this: `tox -e py39` (python 3.9). If you want to run linting: `tox -e lint`.

jadchaar commented:
> I can attempt to make additional pytests once over the main hurdles.
> I don't want the filing-submission.txt so I am putting in another switch like you had for the details.
> Since I did that, I also put one in there for the XBRL data too. Gives the user the option for all three.
> I did make it so the defaults still exhibit the same behavior I found the code in to begin with.

Seems like the most intuitive solution. Maintaining backwards compatibility would be great, but I am wondering if changing the API to something like `dl.get_xbrl` and `dl.get_full_submission` might be a possibility as well. What do you think? Do you think separate functions might be more intuitive than boolean kwargs/switches? We can easily make this a v5.0 release and break the API if we need to.

jreed1701 commented:
> > I can attempt to make additional pytests once over the main hurdles.
> > I don't want the filing-submission.txt so I am putting in another switch like you had for the details.
> > Since I did that, I also put one in there for the XBRL data too. Gives the user the option for all three.
> > I did make it so the defaults still exhibit the same behavior I found the code in to begin with.
> 
> Seems like the most intuitive solution. Maintaining backwards compatibility would be great, but I am wondering if changing the API to something like `dl.get_xbrl` and `dl.get_full_submission` might be a possibility as well. What do you think? Do you think separate functions might be more intuitive than boolean kwargs/switches? We can easily make this a v5.0 release and break the API if we need to.

My opinion follows path of least resistance, but I understand your point. Breaking it out might help new users understand more intuitively as you suggest, however, I recommend implementing as a wrapper that calls the existing function so you can do both. 

I don't plan to make additional functions since the kwargs maintain backward compatibility and do work as the code stands.   If you'd like to push it off to separate PR/Issue I'd consider working it at a later date. You're repo, you're call.  

And thank you for the tox tips. I'm no stranger to pytest but this is fun first time playing with tox. I see the appeal of using it when cross platform compatibility counts.  With respect, a little overkill, but cool nonetheless.
